iree/tools/android/run_module_app/build_apk.sh

```diff
@@ -0,0 +1,296 @@
+
+# IREE project source root.
+IREE_SOURCE_ROOT=$(git rev-parse --show-toplevel)
+# iree-run-module Android app source root.
+IREE_NATIVE_APP_SOURCE_ROOT=${IREE_SOURCE_ROOT}/iree/tools/android/run_module_app
+
+# Directory for IREE native code intermediate intermediate artifacts.
+IREE_NATIVE_LIB_BUILD_DIR=${IREE_ARTIFACT_ROOT}/iree/${IREE_BUILD_TYPE}
+# Directory for IREE native libraries.
+IREE_NATIVE_LIB_DIR=${IREE_ARTIFACT_ROOT}/libs/lib/${IREE_ANDROID_ABI}
+# Directory for Android app assets.
+IREE_ASSET_DIR=${IREE_ARTIFACT_ROOT}/assets
+# Directory for Android app R.class.
+IREE_RESOURCE_GEN_DIR=${IREE_ARTIFACT_ROOT}/rclass
+
+#########################
+# Android build toolchain
+#########################
+
+ANDROID_SDK_BUILD_TOOLS_DIR=${ANDROID_SDK_ROOT}/build-tools/${ANDROID_SDK_BUILD_TOOLS_VERSION}
+ANDROID_SDK_PLATFORMS_DIR=${ANDROID_SDK_ROOT}/platforms/android-${IREE_ANDROID_API_LEVEL}
+
+AAPT_BIN=${ANDROID_SDK_BUILD_TOOLS_DIR}/aapt
+DX_BIN=${ANDROID_SDK_BUILD_TOOLS_DIR}/dx
+ZIPALIGN_BIN=${ANDROID_SDK_BUILD_TOOLS_DIR}/zipalign
+APKSIGNER_BIN=${ANDROID_SDK_BUILD_TOOLS_DIR}/apksigner
+
+if [[ ! -x ${AAPT_BIN} ]] || [[ ! -x ${DX_BIN} ]] || [[ ! -x ${ZIPALIGN_BIN} ]] || [[ ! -x ${APKSIGNER_BIN} ]]; then
+  if [[ -z "${ANDROID_SDK_BUILD_TOOLS_VERSION}" ]]; then
+    echo "Error: 'ANDROID_SDK_BUILD_TOOLS_VERSION' environment variable not set" &>2
+  else
+    echo "Error: '${ANDROID_SDK_BUILD_TOOLS_DIR}' does not point to a valid Android SDK build tools directory containing aapt/apksigner/dx/zipalign" &>2
+  fi
+  exit 1
+fi
+
+AAPT_ADD="${AAPT_BIN} add"
+# Link in the Android framework classes and disable compression for IREE
+# bytecode modules. This allows us to mmap the file directly.
+AAPT_PACK="${AAPT_BIN} package -f -I ${ANDROID_SDK_PLATFORMS_DIR}/android.jar -0 vmfb"
+DX="${DX_BIN} --dex"
+ZIPALIGN="${ZIPALIGN_BIN} -f -p 4"
+APKSIGN="${APKSIGNER_BIN} sign"
+
+JAVAC="${JAVAC_BIN} -classpath ${ANDROID_SDK_PLATFORMS_DIR}/android.jar -sourcepath ${IREE_RESOURCE_GEN_DIR} -d ${IREE_ARTIFACT_ROOT}"
+
+#############################
+# Build IREE native libraries
+#############################
+
+if [[ ! -d ${IREE_NATIVE_LIB_BUILD_DIR} ]]; then
+  mkdir -p ${IREE_NATIVE_LIB_BUILD_DIR}
+fi
+
+echo ">>> Building IREE native libraries <<<"
+
+IREE_NATIVE_LIB_NAME=iree_tools_android_run_module_app_iree_run_module_app
+
+pushd ${IREE_NATIVE_LIB_BUILD_DIR}
+"${CMAKE_BIN}" "${IREE_SOURCE_ROOT}" -G Ninja \
+  -DCMAKE_BUILD_TYPE=${IREE_BUILD_TYPE} \
+  -DCMAKE_TOOLCHAIN_FILE="${ANDROID_NDK_ROOT}/build/cmake/android.toolchain.cmake" \
+  -DANDROID_ABI=${IREE_ANDROID_ABI} \
+  -DANDROID_PLATFORM=android-${IREE_ANDROID_API_LEVEL} \
+  -DIREE_HOST_C_COMPILER="${CC_BIN}" \
+  -DIREE_HOST_CXX_COMPILER="${CXX_BIN}" \
+  -DIREE_BUILD_COMPILER=OFF \
+  -DIREE_BUILD_TESTS=OFF \
+  -DIREE_BUILD_SAMPLES=OFF
+${NINJA_BIN} ${IREE_NATIVE_LIB_NAME}
+popd
+
+#####################
+# Package Android app
+#####################
+
+rm -rf ${IREE_ARTIFACT_ROOT}/*.apk
```

Comment:
Hey @GMNGeoffrey, can you also take a look at these bash commands? Passing env vars to `rm -rf` scares me... isn't there a way to error out if the variable is empty (`${IREE_ARTIFACT_ROOT?}`?)

Reply:
Yes. Please use question mark and quote the expansion.

```suggestion
rm -rf "${IREE_ARTIFACT_ROOT?}"/*.apk
```

You can't quote the `*` or the glob doesn't get expanded though

Reply:
Done.

Reply:
Also I created another intermediate directory to avoid excessive `rm -rf` .